akka-projection-testkit/src/test/java/akka/projection/testkit/javadsl/ProjectionTestKitTest.java

```diff
@@ -0,0 +1,211 @@
+
+        TestProjection prj = new TestProjection(failingSource, strBuffer, i ->  i <= 4);
+
+        try {
+            projectionTestKit.run(prj, () -> assertEquals(strBuffer.toString(), "1-2-3-4"));
+            Assert.fail("should not reach that line");
+        } catch (RuntimeException ex) {
+            assertEquals(ex.getMessage(), streamFailureMsg);
+        }
+    }
+
+    @Test
+    public void runAProjectionWithATestSink() {
+
+        StringBuffer strBuffer = new StringBuffer("");
+        List<Integer> elements = IntStream.rangeClosed(1, 5)
+                .boxed().collect(Collectors.toList());
+
+        TestProjection prj = new TestProjection(Source.from(elements), strBuffer, i -> i <= 5);
+        TestSubscriber.Probe<Done> sinkProbe = projectionTestKit.runWithTestSink(prj);
+
+        sinkProbe.request(5);
+        sinkProbe.expectNextN(5);
+        sinkProbe.expectComplete();
+
+        assertEquals(strBuffer.toString(), "1-2-3-4-5");
+
+    }
+
+
+    private class TestProjection implements Projection<Integer> {
+
+        final private Source<Integer, NotUsed> src;
+        final private StringBuffer strBuffer;
+        final private Predicate<Integer> predicate;
+        final private SharedKillSwitch killSwitch;
+
+        final private Promise<Done> promiseToStop;
+
+        private TestProjection(Source<Integer, NotUsed> src, StringBuffer strBuffer, Predicate<Integer> predicate) {
+            this.src = src;
+            this.strBuffer = strBuffer;
+            this.predicate = predicate;
+            this.killSwitch = KillSwitches.shared(projectionId().id());
+            this.promiseToStop = Promise.apply();
+        }
+
+
+        @Override
+        public ProjectionId projectionId() {
+            return ProjectionId.of("test-projection", "00");
+        }
+
+        @Override
+        public akka.stream.scaladsl.Source<Done, NotUsed> mappedSource(ClassicActorSystemProvider systemProvider) {
+            Source<Integer, NotUsed> via = src.via(killSwitch.flow());
+            return src.via(killSwitch.flow())
+                    .mapAsync(1, (Function<Integer, CompletionStage<Done>>) this::process).asScala();
+        }
+
+        private CompletionStage<Done> process(Integer i) {
+            if (predicate.test(i)) concat(i);
+            return CompletableFuture.completedFuture(Done.getInstance());
+        }
+
+        StringBuffer concat(Integer i) {
+            if (strBuffer.toString().isEmpty())
+                return strBuffer.append(i);
+            else
+                return strBuffer.append("-").append(i);
+        }
+
+        @Override
+        public void run(ClassicActorSystemProvider systemProvider) {
+            CompletionStage<Done> done = mappedSource(systemProvider).asJava().runWith(Sink.ignore(), systemProvider);
+            promiseToStop.completeWith(FutureConverters.toScala(done));
+        }
+
+        @Override
+        public Future<Done> stop(ExecutionContext ec) {
```

Comment:
`Projection.stop` is returning Scala Future, but I believe we will remove the start and stop methods as they are not really needed at API level. So for the moment I'm leaving it as is. 

We may need to have a `scaladsl.Projection` and a `javadsl.Projection` for the `mappedSource`, but this still need to be explored on a future PR.